File: Sources/QTube/Services/LinkExtractor.swift
import Foundation

enum LinkExtractor {
    private static let acceptedHosts = [
        "youtube.com",
        "www.youtube.com",
        "m.youtube.com",
        "music.youtube.com",
        "youtu.be"
    ]

    private static let detector = try? NSDataDetector(types: NSTextCheckingResult.CheckingType.link.rawValue)

    static func youtubeURLs(from text: String) -> [URL] {
        guard !text.isEmpty else { return [] }

        let range = NSRange(text.startIndex..<text.endIndex, in: text)
        let matches = detector?.matches(in: text, options: [], range: range) ?? []
        var seen = Set<String>()

        return matches.compactMap { match in
            guard let url = match.url, isYouTubeURL(url) else { return nil }

            let normalized = url.absoluteString
            guard seen.insert(normalized).inserted else { return nil }
            return url
        }
    }

    static func youtubeURL(from text: String) -> URL? {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty,
              !trimmed.contains(where: { $0.isWhitespace }),
              let url = URL(string: trimmed),
              isYouTubeURL(url)
        else { return nil }
        return url
    }

    static func playlistID(from url: URL) -> String? {
        guard let components = URLComponents(url: url, resolvingAgainstBaseURL: false),
              let queryItems = components.queryItems else { return nil }
        guard let item = queryItems.first(where: { $0.name.lowercased() == "list" }),
              let value = item.value, !value.isEmpty else { return nil }
        return value
    }

    static func isPlaylistURL(_ url: URL) -> Bool {
        guard isYouTubeURL(url) else { return false }
        if playlistID(from: url) != nil {
            return true
        }
        let path = url.path.lowercased()
        return path.hasPrefix("/@") || path.hasPrefix("/channel/") || path.hasPrefix("/c/") || path.hasPrefix("/user/")
    }

    static func isYouTubeMix(listID: String) -> Bool {
        listID.hasPrefix("RD") || listID.hasPrefix("UL") || listID.hasPrefix("PU")
    }

    static func isYouTubeMixURL(_ url: URL) -> Bool {
        guard let listID = playlistID(from: url) else { return false }
        return isYouTubeMix(listID: listID)
    }

    static func isWatchVideoWithPlaylistURL(_ url: URL) -> Bool {
        guard isYouTubeURL(url), let listID = playlistID(from: url), !listID.isEmpty else { return false }
        let path = url.path.lowercased()
        return path.contains("/watch") || url.host?.lowercased() == "youtu.be"
    }

    static func cleanVideoURL(from url: URL) -> URL {
        guard var components = URLComponents(url: url, resolvingAgainstBaseURL: false) else { return url }
        components.queryItems = components.queryItems?.filter { item in
            let name = item.name.lowercased()
            return name != "list" && name != "index" && name != "start_radio"
        }
        return components.url ?? url
    }

    static func playlistURL(from listID: String) -> URL? {
        URL(string: "https://www.youtube.com/playlist?list=\(listID)")
    }

    private static func isYouTubeURL(_ url: URL) -> Bool {
        guard let scheme = url.scheme?.lowercased(),
              scheme == "http" || scheme == "https",
              let host = url.host?.lowercased()
        else { return false }

        return acceptedHosts.contains(host) || host.hasSuffix(".youtube.com")
    }
}

File: windows/QTube.Windows/Services/LinkExtractor.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;
using System.Web;

namespace QTube.Windows.Services
{
    public static class LinkExtractor
    {
        private static readonly string[] AcceptedHosts = new[]
        {
            "youtube.com",
            "www.youtube.com",
            "m.youtube.com",
            "music.youtube.com",
            "youtu.be"
        };

        private static readonly Regex UrlRegex = new Regex(
            @"(https?://[^\s<>""]+)",
            RegexOptions.Compiled | RegexOptions.IgnoreCase
        );

        public static List<Uri> ExtractYouTubeUrls(string text)
        {
            if (string.IsNullOrWhiteSpace(text)) return new List<Uri>();

            var matches = UrlRegex.Matches(text);
            var seen = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
            var results = new List<Uri>();

            foreach (Match match in matches)
            {
                if (Uri.TryCreate(match.Value, UriKind.Absolute, out var uri) && IsYouTubeUrl(uri))
                {
                    if (seen.Add(uri.AbsoluteUri))
                    {
                        results.Add(uri);
                    }
                }
            }

            return results;
        }

        public static Uri? ParseSingleYouTubeUrl(string text)
        {
            if (string.IsNullOrWhiteSpace(text)) return null;
            var trimmed = text.Trim();
            if (trimmed.Any(char.IsWhiteSpace)) return null;

            if (Uri.TryCreate(trimmed, UriKind.Absolute, out var uri) && IsYouTubeUrl(uri))
            {
                return uri;
            }
            return null;
        }

        public static string? GetPlaylistId(Uri url)
        {
            if (!IsYouTubeUrl(url)) return null;
            var query = HttpUtility.ParseQueryString(url.Query);
            var list = query["list"];
            return string.IsNullOrWhiteSpace(list) ? null : list;
        }

        public static bool IsPlaylistUrl(Uri url)
        {
            if (!IsYouTubeUrl(url)) return false;
            if (!string.IsNullOrEmpty(GetPlaylistId(url))) return true;

            string path = url.AbsolutePath.ToLowerInvariant();
            return path.StartsWith("/@") || path.StartsWith("/channel/") || path.StartsWith("/c/") || path.StartsWith("/user/");
        }

        public static bool IsYouTubeMix(string listId)
        {
            return listId.StartsWith("RD", StringComparison.OrdinalIgnoreCase) ||
                   listId.StartsWith("UL", StringComparison.OrdinalIgnoreCase) ||
                   listId.StartsWith("PU", StringComparison.OrdinalIgnoreCase);
        }

        public static bool IsYouTubeMixUrl(Uri url)
        {
            var listId = GetPlaylistId(url);
            return listId != null && IsYouTubeMix(listId);
        }

        public static bool IsWatchVideoWithPlaylistUrl(Uri url)
        {
            if (!IsYouTubeUrl(url)) return false;
            var listId = GetPlaylistId(url);
            if (string.IsNullOrEmpty(listId)) return false;

            string path = url.AbsolutePath.ToLowerInvariant();
            return path.Contains("/watch") || string.Equals(url.Host, "youtu.be", StringComparison.OrdinalIgnoreCase);
        }

        public static Uri CleanVideoUrl(Uri url)
        {
            var builder = new UriBuilder(url);
            var query = HttpUtility.ParseQueryString(builder.Query);

            query.Remove("list");
            query.Remove("index");
            query.Remove("start_radio");

            builder.Query = query.ToString();
            return builder.Uri;
        }

        public static Uri? PlaylistUrlFromId(string listId)
        {
            return new Uri($"https://www.youtube.com/playlist?list={listId}");
        }

        public static bool IsYouTubeUrl(Uri url)
        {
            if (url.Scheme != Uri.UriSchemeHttp && url.Scheme != Uri.UriSchemeHttps) return false;
            string host = url.Host.ToLowerInvariant();

            return AcceptedHosts.Contains(host) || host.EndsWith(".youtube.com", StringComparison.OrdinalIgnoreCase);
        }
    }
}
